Fixes potential PoolState.Waiting overflow and includes some minor changes

NinoFloris opened the pull request:
Changes the `PoolState.Waiting` short to an int and adds a checked increment (one branch instruction that'll always be branch predicted away). There's a test that asserts whether it's throwing OverflowException correctly.

I've also added an assert to create a dependency for PoolSizeLimit on the max value allowed by the state idle and busy primitives. I don't expect it to ever be valuable as a runtime test but it's a good case to be reminded of if this limit should become customizable.

The cancellation token commit feels more consistent with the intent of the work being done whether it's being awaited asynchronously or not but I'm entirely ambivalent either way it goes.

I also did take a look at any other potential places to optimize but it's pretty trimmed!

P.S. 'asking for a friend'
How easy would it be to add another layer of abstraction between the actual pg connection and the pool so we could multiplex work?

YohDeadfall commented:
> How easy would it be to add another layer of abstraction between the actual pg connection and the pool so we could multiplex work?

@roji is working on it (see [his branches](https://github.com/roji/Npgsql/branches/all?utf8=%E2%9C%93&query=multiplex)) and there is #1982 to track progress on it.

YohDeadfall commented on src/Npgsql/ConnectorPool.cs:
```diff
@@ -260,7 +263,7 @@ internal async ValueTask<NpgsqlConnector> AllocateLong(NpgsqlConnection conn, Np
                     if (_pruningTimer == null && newState.Total > _min)
                     {
-                        var newPruningTimer = new Timer(PruneIdleConnectors);
+                        var newPruningTimer = new Timer(s => PruneIdleConnectors(s));
```
Any reason to do this? In that case you're creating a new method inside an internal class which internally calls `PruneIdleConnectors`.

NinoFloris replied:
You're right, long day, it actually still captures the this reference. Thought it was a static method. Where what I did actually makes sense as C# doesn't cache any method group conversions but it does cache non capturing lambdas. 

Let me do a revision of that commit as I see a few other places like it to save allocations.

YohDeadfall commented on src/Npgsql/ConnectorPool.cs:
```diff
@@ -329,11 +336,11 @@ await Task.WhenAny(tcs.Task, Task.Delay(timeLeft, delayCancellationToken.Token))
                                 {
                                     var timeLeft = timeout.TimeLeft;
-                                    if (timeLeft <= TimeSpan.Zero || !tcs.Task.Wait(timeLeft))
+                                    if (timeLeft <= TimeSpan.Zero || !tcs.Task.Wait((int)timeLeft.TotalMilliseconds, cancellationToken))
```
What is a reason to get milliseconds and convert them to `int`? If the value is larger than `int.MaxValue` there could be no exception at all because it could be trimmed.

NinoFloris replied:
This is the maximum expected in Task.Wait, it would be good to indeed add a argument exception like the overload taking just a TimeSpan does, will do.
```
public bool Wait(TimeSpan timeout)
{
      long totalMilliseconds = (long) timeout.TotalMilliseconds;
      if (totalMilliseconds < -1L || totalMilliseconds > (long) int.MaxValue)
        ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.timeout);
      return this.Wait((int) totalMilliseconds, new CancellationToken());
}
```
There's just no overload taking a TimeSpan and a CancellationToken

NinoFloris replied:
To add this could have already happened before with an insanely high NpgsqlTimeout, it's just made visible now.

YohDeadfall replied:
> There's just no overload taking a TimeSpan and a CancellationToken

Then it's a good idea to open a pull request in Core FX repo and add it to .NET Standard 2.1 too.

NinoFloris replied:
Done

YohDeadfall commented on src/Npgsql/ConnectorPool.cs:
```diff
@@ -111,6 +115,7 @@ internal ConnectorPool(NpgsqlConnectionStringBuilder settings, string connString
                 : settings.ToStringWithoutPassword();
 
             _pruningInterval = TimeSpan.FromSeconds(Settings.ConnectionPruningInterval);
+            _pruningTimerCallback = PruneIdleConnectors;
```
This could be made even better. Just change `PruneIdleConnectors` to be static and pass the current instance to the timer on line 268. So there will be only one instance of the delegate pointing to the method.

NinoFloris replied:
Done, it doesn't win any beauty contests but on perf it's optimal :)

roji commented:
Thanks for the PR @NinoFloris :)

>> How easy would it be to add another layer of abstraction between the actual pg connection and the pool so we could multiplex work?
>
> @roji is working on it (see his branches) and there is #1982 to track progress on it.

Unfortunately I'm a bit stalled on that because of other work (mainly EF)... Will hopefully get some time to work on it at some point. FWIW multiplexing wouldn't necessary be another layer of abstraction, it could be more invasive than that. We'll see.

Will probably give this PR a review during the weekend.

NinoFloris commented:
> Unfortunately I'm a bit stalled on that because of other work (mainly EF)...

Understood ;) and obviously take all the time you need.

YohDeadfall commented on src/Npgsql/ConnectorPool.cs:
```diff
@@ -84,9 +84,10 @@ public override string ToString()
         Timer? _pruningTimer;
         readonly TimeSpan _pruningInterval;
-
+        
```
Remove whitespaces from this line.

YohDeadfall replied:
Whitespaces are still here. Have you pushed changes?

YohDeadfall commented on src/Npgsql/ConnectorPool.cs:
```diff
@@ -526,21 +536,24 @@ void CloseConnector(NpgsqlConnector connector, bool wasIdle)
-            for (var i = 0; i < _idle.Length; i++)
+            var idleLifetime = pool.Settings.ConnectionIdleLifetime;
+                
+            
```
Remove one of these two empty lines.

YohDeadfall replied:
Looks like the IDE is poking fun at you.

YohDeadfall commented on test/Npgsql.Tests/PoolTests.cs:
```diff
@@ -138,6 +138,36 @@ public async Task TimeoutGettingConnectorFromExhaustedPoolAsync()
+        [Test]
+        public void OverflowExceptionWhenTooManyWaiting()
+        {
+            var connString = new NpgsqlConnectionStringBuilder(ConnectionString) {
```
nit: We put curly brackets on the next line in such cases.

NinoFloris commented:
@YohDeadfall addressed the nits, some friendly feedback though. Proper consistency makes contributing in the expected conventions a breeze, but nits like these on PRs will always be a big agitant by their very nature, especially if they're the first or last thing people have to do. 

In this case, some of the nits you bring forward aren't consistently correct in the codebase or the same file even. That makes people feel picked on while 'special people' can just bypass the 'hassle' (please understand that this is a simplified version of the nuanced view I'm trying to convey). I deeply understand how these situations emerge but its something to reflect on, aligning 'what we expect' and 'what the maintainers do' is crucial to a welcoming and fair atmosphere.

If it's this important I would personally look for some automation to do a pass as part of PRs. C# really is quite amenable to automatic formatting compared to say F#.

roji commented on src/Npgsql/ConnectorPool.cs:
```diff
@@ -98,6 +99,9 @@ public override string ToString()
 
         internal ConnectorPool(NpgsqlConnectionStringBuilder settings, string connString)
         {
+            Debug.Assert(PoolSizeLimit <= short.MaxValue,
```
Just a thought - we could make this a runtime check - there's no perf impact whatsoever (just a setup check), and really confused users may accidentally set this number very high. It's true that they'll like get errors from the PostgreSQL about surpassing max_connections, but for correctness why not do it anyway.

NinoFloris replied:
I don't get what you mean with runtime check? The assert is on an internal const.

roji commented on src/Npgsql/ConnectorPool.cs:
```diff
@@ -320,20 +328,22 @@ await Task.WhenAny(tcs.Task, Task.Delay(timeLeft, delayCancellationToken.Token))
-                                    var timeLeft = timeout.TimeLeft;
-                                    if (timeLeft <= TimeSpan.Zero || !tcs.Task.Wait(timeLeft))
+                                    var timeLeft = timeout.TimeLeft.TotalMilliseconds;
+                                    if (timeLeft > int.MaxValue)
```
Thanks... So if my calculations are correct we only support around 24 days of timeout - let's see if someone complains :)

roji commented on src/Npgsql/ConnectorPool.cs:
```diff
@@ -320,20 +328,22 @@ await Task.WhenAny(tcs.Task, Task.Delay(timeLeft, delayCancellationToken.Token))
                                 }
                                 else
-                                    tcs.Task.Wait();
+                                    tcs.Task.Wait(cancellationToken);
```
You wrote:

> The cancellation token commit feels more consistent with the intent of the work being done whether it's being awaited asynchronously or not but I'm entirely ambivalent either way it goes.

Yeah, but unless I'm mistaken there's no way this cancellation token can ever trigger here in the sync path:
* We could not have gotten a ctoken from the user, and
* Timeout isn't implemented via a ctoken as in the async path

If that's the case and I haven't missed anything, then I think it's better not to pass the ctoken in to make the code more clear.

NinoFloris replied:
I don't get what you mean with runtime check? The assert is on an internal const.

NinoFloris replied:
Not in your implementation, I usually like to make coherent apis that when I generalize them to be used in multiple projects there is usually only api shape work left. I can revert that commit of course, but it doesn't do any harm right now, it all shortcircuits on `token.CanBeCanceled` if it's always a default token.

roji replied:
I'd rather we remove it. When reading the code it gives the mistaken impression that cancellation is possible there (i.e. I'm not worried about perf or anything, just readability in this already pretty complicated component).

NinoFloris replied:
I'd say that actually improves maintainability "cancellation is always possible" no exceptions based on branching. But I'll revert the commit :)

roji replied:
Thanks, I guess it's a matter of taste/opinion - "cancellation is always possible", but not in this actual specific case. When I saw your commit adding the token, I spend a bit of time analyzing the code to try to understand in which case it may trigger - I'd like to save the mental effort from myself or anyone else reading that code.

Anyway thanks for reverting!

roji commented on src/Npgsql/ConnectorPool.cs:
```diff
@@ -305,7 +313,7 @@ internal async ValueTask<NpgsqlConnector> AllocateLong(NpgsqlConnection conn, Np
                                     // do complete successfully
                                     var delayCancellationToken = new CancellationTokenSource();
-                                    using (cancellationToken.Register(() => delayCancellationToken.Cancel()))
+                                    using (cancellationToken.Register(s => ((CancellationTokenSource)s).Cancel(), delayCancellationToken))
```
Great!

roji commented:
@NinoFloris regarding your comment on the nits, I get what you're saying and it definitely merits thinking. In general I try to express myself in a way that wouldn't discourage contributors (e.g. accompany the nits with positive feedback and explicitly downplay their importance). I don't have the impression that contributors have had a negative experience contributing here (I certainly hope you didn't!), but then I may not know about it if it happened.

Re automated PR tools, it's not a bad idea but we have very little extra time for this kind of work - in my experience it takes quite a bit of time to get it exactly the way you want it.

YohDeadfall commented:
@NinoFloris, first I would like to sorry if my responses hurt you or make you feel uncomfortable. There is nothing personal, and all contributors are welcome.


Speaking about nits, some of code in this repo is old enough and could be improperly formatted. So we ask contributors to fix them before accepting changes, and we do the same thing too when detect them in code we are touching.

austindrenski commented on test/Npgsql.Tests/PoolTests.cs:
```diff
@@ -138,6 +138,36 @@ public async Task TimeoutGettingConnectorFromExhaustedPoolAsync()
+                finally
+                {
+                    // Restore state for the closes work correctly.
+                    pool!.State = state; 
```
Out of curiosity, what was the behavior without the `finally`? Was the dispose path throwing? Or is this just here to cleanup the bad pool?

NinoFloris replied:
Cleanup the bad pool. If you don't the pool implementation tries to hand off a connection to a waiting tcs that isn't there and it thinks it raced the enqueue (as the change is waiting++ but also enqueue) so it loops forever.

austindrenski commented on src/Npgsql/ConnectorPool.cs:
```diff
@@ -320,20 +328,22 @@ await Task.WhenAny(tcs.Task, Task.Delay(timeLeft, delayCancellationToken.Token))
-                                    if (timeLeft <= TimeSpan.Zero || !tcs.Task.Wait(timeLeft))
+                                    var timeLeft = timeout.TimeLeft.TotalMilliseconds;
+                                    if (timeLeft > int.MaxValue)
+                                        throw new ArgumentOutOfRangeException(nameof(timeout), "timeout.TimeLeft.TotalMilliseconds must be smaller than int.MaxValue.");
```
Just a thought, but since `PoolTests` already has little tests for things like `MaxPoolSize < MinPoolSize`, it might be nice to add one for this too.

E.g.:

https://github.com/npgsql/npgsql/blob/d78f3d6c8ca88d6adb6fb26d77bd8d8293269d0c/src/Npgsql/ConnectorPool.cs#L101-L102


https://github.com/npgsql/npgsql/blob/d78f3d6c8ca88d6adb6fb26d77bd8d8293269d0c/test/Npgsql.Tests/PoolTests.cs#L39-L44

austindrenski replied:
Same for `PoolSizeLimit <= short.MaxValue`, if you change it to a runtime check.

NinoFloris replied:
I'm not going to add a test for a timeout that can take 24 days and is artificially constrained by the BCLs choice for a primitive type. 🙃

austindrenski replied:
Ope! Yeahhh, I didn't think that one through on my end!

NinoFloris commented:
@YohDeadfall I know that's why I wrote my reply like that, giving feedback over the internet sucks vs face to face sucks but it's the best I can do.

YohDeadfall commented on test/Npgsql.Tests/PoolTests.cs:
```diff
@@ -337,7 +374,8 @@ public void PhysicalOpenFailure()
         {
-            var connString = new NpgsqlConnectionStringBuilder(ConnectionString) {
+            var connString = new NpgsqlConnectionStringBuilder(ConnectionString) 
+            {
```
Thank you for fixing this too!

NinoFloris replied:
No problem!

roji commented:
Merged and backported to 4.0.8 via 7b6c0d523daf5ace7a0455fd45c67803ea683535.

Thanks @NinoFloris! Curious to see what you come up with next :)
